selectors/selectors.md

```diff
@@ -302,67 +265,152 @@ If it's a file in some directory, you can also start at a deeper level:
+type ExploreRecursiveEdge struct {}
+
+## ExploreUnion allows selection to continue with two or more distinct selectors
+## while exploring the same tree of data.
+##
+## ExploreUnion can be used to apply a Matcher on one node (causing it to
+## be considered part of a (possibly labelled) result set), while simultaneously
+## continuing to explore deeper parts of the tree with another selector,
+## for example.
+type ExploreUnion list [Selector]
+
+## Note that ExploreConditional versus a Matcher with a Condition are distinct:
+## ExploreConditional progresses deeper into a tree;
+## whereas a Matcher with a Condition may look deeper to make its decision,
+## but returns a match for the node it's on rather any of the deeper values.
+type ExploreConditional struct {
+	condition Condition (alias "&")
+	next Selector (alias ">")
 }
 
-# SelectRecursive applies a selector to this node, then in addition traverses
-# to any nodes matched and applies itself again to that node.  A depth counter
-# is decremented with each recursion to ensure the process terminates.
-type SelectRecursive struct {
-	depthLimit Int (alias "d")
-	next Selector (alias ":")
-	cidLimit Link
+## Matcher marks a node to be included in the "result" set.
+## (All nodes traversed by a selector are in the "covered" set (which is a.k.a.
+## "the merkle proof"); the "result" set is a subset of the "covered" set.)
+##
+## In libraries using selectors, the "result" set is typically provided to
+## some user-specified callback.
+##
+## A selector tree with only "explore*"-type selectors and no Matcher selectors
+## is valid; it will just generate a "covered" set of nodes and no "result" set.
+type Matcher struct {
+	onlyIf optional Condition # match is true based on position alone if this is not set.
+	label optional String # labels can be used to match multiple different structures in one selection.
 }
 
-# SelectUnion is a union of other selectors: if any of the selectors says
-# a node matches, it matches; and if any of the selectors return a selector
-# for a pathsegment, it'll be used; and if more than one selector returns
-# a selector for a pathsegment, a new union selector will be implicitly
-# generated during evaluation to carry all of them through.
-type SelectUnion list [Selector]
+## Condition is expresses a predicate with a boolean result.
+##
+## Condition clauses are used several places:
+##   - in Matcher, to determine if a node is selected.
+##   - in ExploreRecursive, to halt exploration.
+##   - in ExploreConditional,
+##
+##
+## TODO -- Condition is very skeletal and incomplete.
+## The place where Condition appears in other structs is correct;
+## the rest of the details inside it are not final nor even completely drafted.
+type Condition union {
+	# We can come back to this and expand it later...
+	# TODO: figure out how to make this recurse correctly, so I can say "hasField{hasField{or{hasValue{1}, hasValue{2}}}}".
```

Comment:
It seems like we're heading towards matching / explore being equal in capability, and I wonder if that's neccesary. Maybe there is only matching at the end -- i.e. exploreFields{exploreFields{or{hasValue{1}, hasValue{2}} -- because that's the equivalent yes? -- I would argue against having matching be able to recurse

Reply:
Yeah, it's possible that it would turn out wiser to limit this feature.

The one major case I can think of that's super super common and definitely needs to be supported is: given some nodes in an array, each is a map, and I want to decide whether or not to explore or match that node based on a property in it.

I don't have any example use cases of deeper recursion.

Another possibly sensible limit might be to say "this can recurse, but not across another link load boundary"?

Looking at what `jq` does might also be interesting.  That system is subject to different constraints than we are (namely, link loading doesn't exist for them, so of course it can't be a performance cost issue), but if there *are* examples we can find of `jq` usage in the wild that do any deeper conditions, that might be interesting to consider.

Reply:
Marking this thread as resolved because it's a known todo and it's staying that way until future iterations -- not gonna handle in this PR.